Facilitate use of TFPs with (v,e) variable set

GiudGiud opened the pull request:
The log grid in e should help with those simulations having e reach from 1e5 to 1e12
and the grid output will help debug the conversions

Some parameters were being ignored. They are now explicitly called out
refs #28020

GiudGiud committed:
Add options for log-grid in internal energy
refs #28020

GiudGiud committed:
Add output of interpolation grids for pt from ve
refs #28020

GiudGiud committed:
Add testing for log_e feature
refs #28020

GiudGiud committed:
Correct source of bounds info in OOB logs
Add warning in constructor for these parameters

GiudGiud committed:
Add testing for (v,e) tabulation output feature

GiudGiud commented:
@licharlot 

this does not put us very far from being able to interpolate everything directly from (v,e), instead of using (v,e) -> (p,T) -> property we care about

simopier reviewed (commented):
Thank you for this, @GiudGiud!

simopier commented on modules/fluid_properties/src/fluidproperties/TabulatedBicubicFluidProperties.C:
```diff
@@ -419,9 +437,17 @@ TabulatedBicubicFluidProperties::outputWarnings(unsigned int num_nans_p,
+  {
     warning_message += p_oob;
+    warning_message += ("Pressure bounds " + source + ": [" + std::to_string(_pressure_min) + ", " +
+                        std::to_string(_pressure_max) + "]");
```
```suggestion
                        std::to_string(_pressure_max) + "]\n");
```

simopier commented on modules/fluid_properties/src/fluidproperties/TabulatedBicubicFluidProperties.C:
```diff
@@ -419,9 +437,17 @@ TabulatedBicubicFluidProperties::outputWarnings(unsigned int num_nans_p,
+  {
     warning_message += T_oob;
+    warning_message += ("Temperature bounds " + source + ": [" + std::to_string(_temperature_min) +
+                        ", " + std::to_string(_temperature_max) + "]");
```
```suggestion
                        ", " + std::to_string(_temperature_max) + "]\n");
```

simopier reviewed (commented):
To clean up the console.

GiudGiud commented:
Thanks for the review. I ll batch these with Josh or Lise's comments

GiudGiud committed:
Add other property output in (v,e) table

GiudGiud committed:
Clean up VE FP tests
the log_e test is not ideal as it is since it is out of bounds with e_co2 = -30000
This is not an unexpected condition for the object though

simopier commented on modules/fluid_properties/test/tests/tabulated/gold/fluid_properties_out_ve.csv:
```diff
@@ -1,101 +1,101 @@
-1.13323, 149156, 100000, 565.022
-1.13323, 179821, 100000, 600
+specific_volume, internal_energy, pressure, temperature, density, enthalpy, viscosity, k, c, cv, cp, entropy
+0.0137352, -96165.8, 3e+06, 28072.8057, -54960.3, 1.47169e-05, 0.0179746, 228.551, 777.464, 1342.27, -797.032, 
```
Some of the temperatures go super high here.

GiudGiud replied:
not much I can do about it. ranges in v, e are brutally large, the pressure is stuck at the max value

simopier commented on modules/fluid_properties/test/tests/tabulated/gold/fluid_properties_out_ve.csv:
```diff
@@ -1,101 +1,101 @@
+0.0137352, 87825.4, 3e+06, 506.05872.8057, 185619, 2.44675e-05, 0.0349016, 340.502, 837.941, 1057.42, -162.621, 
+0.0137352, 118491, 3e+06, 541.2872.8057, 223134, 2.58809e-05, 0.0378051, 352.505, 858.69, 1072.86, -90.9611, 
+0.0137352, 149156, 3e+06, 575.90572.8057, 260550, 2.7234e-05, 0.0406233, 363.75, 878.177, 1088.4, -23.9625, 
+0.0137352, 179821, 3e+06, 60072.8057, 286905, 2.81547e-05, 0.0425552, 371.297, 891.176, 1099.19, 20.868, 
```
High temperature here.

simopier commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -1104,38 +1135,92 @@ TabulatedFluidProperties::s_from_h_p(Real h, Real pressure) const
+              file_out << k_from_v_e(v_val, e_val) << ", ";
+            else if (i == _c_idx)
+              file_out << c_from_v_e(v_val, e_val) << ", ";
+            else if (i == _cp_idx)
+              file_out << cp_from_v_e(v_val, e_val) << ", ";
+            else if (i == _cv_idx)
+              file_out << cv_from_v_e(v_val, e_val) << ", ";
```
```suggestion
            else if (i == _cv_idx)
              file_out << cv_from_v_e(v_val, e_val) << ", ";
            else if (i == _cp_idx)
              file_out << cp_from_v_e(v_val, e_val) << ", ";
```
I think it should be the reverse order.

GiudGiud replied:
the 'if' will take care of that

simopier reviewed (commented):
Just a couple comments and suggestions. Thanks @GiudGiud!

joshuahansel commented on modules/fluid_properties/test/tests/tabulated/gold/fluid_properties_out_ve.csv:
```diff
@@ -0,0 +1,101 @@
+specific_volume, internal_energy, pressure, temperature, density, enthalpy, viscosity, k, c, cv, cp, entropy
+0.0137352, -96165.8, 3e+06, 28072.8057, -54960.3, 1.47169e-05, 0.0179746, 228.551, 777.464, 1342.27, -797.032, 
+0.0137352, -65500.6, 3e+06, 318.82672.8057, -9085.85, 1.6389e-05, 0.020222, 257.834, 739.584, 1090.78, -643.269, 
```
There's a comma at the end of each line. I'm guessing this would confuse some readers to make them think there's another column.

GiudGiud replied:
fixed. pushing in 3 mins

joshuahansel reviewed (changes requested):
Please doc new features

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2304531/) on 8f7b3f5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28021/site)

This comment will be updated on new commits.

GiudGiud committed:
Address Josh's review:
- add documentation for log_e grid
- add documentation for ve table output
- improve output formatting

GiudGiud committed:
Adapt porous flow tests and examples to now more-sane logic on TFP use

GiudGiud committed:
Adapt unit tests to new file logic
Make the indexing of FPs safer than pre-initializing to a valid value, using a value that will cause a seg fault if misused

GiudGiud committed:
Allow an exception for two sources of FP data for porous flow

GiudGiud commented:
Well this was a pain. The weird behavior of "if file exists, use that, if not write it" is used all over the porous flow module.

GiudGiud commented:
Falcon patch is up

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2307401/) on 812f01e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>efc7cb</th>
      <th colspan=4>#28021 812f01</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.06%</td>
      <td>85.08%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104499</td>
      <td>104514</td>
      <td>+15</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18348</td>
      <td>18333</td>
      <td>-15</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28021/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28021/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>efc7cb</th>
      <th colspan=4>#28021 812f01</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.48%</td>
      <td>85.61%</td>
      <td>+0.13%</td>
      <td>90.55%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7266</td>
      <td>7337</td>
      <td>+71</td>
      <td>115</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1234</td>
      <td>1233</td>
      <td>-1</td>
      <td>12</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28021/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28021/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28021/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
arm failure unrelated

GiudGiud commented on modules/porous_flow/examples/ates/ates.i:
```diff
@@ -17,54 +17,54 @@ produce_time = 91 # days
 num_cycles = 5 # Currently needs to be <= 10
 
-cycle_length = ${fparse inject_time + store_time + produce_time + rest_time}
+cycle_length = '${fparse inject_time + store_time + produce_time + rest_time}'
```
yes
